python/cudf/groupby/groupby.py

```diff
@@ -110,6 +112,66 @@ def _apply_basic_agg(self, agg_type, sort_results=False):
         """
         return _cpp_apply_basic_agg(self, agg_type, sort_results=sort_results)
 
+    def apply_multiindex_or_single_index(self, result):
+        if 1 == len(self._by):
+            from cudf.dataframe import index
+            idx = index.as_index(result[self._by[0]])
+            idx.name = self._by[0]
+            result = result.drop(idx.name)
+            if idx.name == self._LEVEL_0_INDEX_NAME:
+                idx.name = None
+            result = result.set_index(idx)
+            return result
+        else:
+            levels = []
+            codes = DataFrame()
+            names = []
+            # Note: This is an O(N^2) solution using gpu masking
+            # to compute new codes for the MultiIndex. There may be
+            # a faster solution that could be executed on gpu at the same
+            # time the groupby is calculated.
+            for by in self._by:
+                level = result[by].unique()
+                code = result[by]
+                for idx, value in enumerate(level):
+                    level_mask = code == value
+                    code = code.masked_assign(idx, level_mask)
+                levels.append(level)
+                codes[by] = code
+                names.append(by)
+            from cudf import MultiIndex
+            multi_index = MultiIndex(levels=levels,
+                                     codes=codes,
+                                     names=names)
+            for by in self._by:
+                result = result.drop(by)
+            return result.set_index(multi_index)
```

Comment:
Will need to do a walkthrough of this code, because for a large number of groups this looks very problematic.

Reply:
I see why. Perhaps instead of doing this, I identify the set of columns not in by and construct a new dataframe using only the positive columns?

Reply:
I changed it to
```
            for col in result.columns:
                if col not in self._by:
                    final_result[col] = result[col]
```